feat: artifact processor extender proposal (#1)
hyy0322/community: proposals/artifact-processor-extender.md

# Proposal: `Artifact Processor Extender`

> Author: `Yiyang Huang @hyy0322`

Reviewer: Add Jian Zhu and Ce Gao here

> Discussion: `Link do discussion issue, if applicable`

Reviewer: ```suggestion
Discussion: [goharbor/harbor#12013](https://github.com/goharbor/harbor/issues/12013)
```

## Abstract

Add support to Harbor for using remote artifact processor to process user defined OCI Artifact.

## Background

> There are three types of artifacts, which is image, CNAB, Helm Chart, supported by Harbor. Each of them implements there own processor to process artifact data. If users want to define a new kind of artifact, they need to add code implementing the processor logic in harbor core service. In this way, users need to maintain their own fork, which is unnecessary if harbor provide ability for user to extende their own processor.

Reviewer: ```suggestion
There are three types of artifacts, which are image, CNAB, Helm Chart, supported by Harbor. Each of them implements its processor to process artifact data. If users want to define a new kind of artifact, they need to implement the processor logic in goharbor/harbor core service, which greatly limits the scalability and extensibility of Harbor.
```

## Proposal

> Harbor now has defined artifact ```Processor``` interface.

Reviewer: ```suggestion
```Processor``` interface is defined in Harbor, which is used to process the specified artifact.
```

```
// Processor processes specified artifact
type Processor interface {
	// GetArtifactType returns the type of one kind of artifact specified by media type
	GetArtifactType() string
	// ListAdditionTypes returns the supported addition types of one kind of artifact specified by media type
	ListAdditionTypes() []string
	// AbstractMetadata abstracts the metadata for the specific artifact type into the artifact model,
	// the metadata can be got from the manifest or other layers referenced by the manifest.
	AbstractMetadata(ctx context.Context, manifest []byte, artifact *artifact.Artifact) error
	// AbstractAddition abstracts the addition of the artifact.
	// The additions are different for different artifacts:
	// build history for image; values.yaml, readme and dependencies for chart, etc
	AbstractAddition(ctx context.Context, artifact *artifact.Artifact, additionType string) (addition *Addition, err error)
}
```

```Registry``` is defined to store ```Processor```

```
var (
	// Registry for registered artifact processors
	Registry = map[string]Processor{}
)
```

> Artifact processor is aim to deal with ```artifact.Artifact``` and return some data related to ```artifact.Artifact```. So we can define a ```ExtenderProcessor``` like this.

Reviewer: ```suggestion
The artifact processor is aimed to deal with ```artifact.Artifact``` and return some data related to ```artifact.Artifact```. Thus we can define an ```ExtenderProcessor``` like this.
```

```ExtenderProcessor``` interface

```

> type ExtenderProcessor interface {

Reviewer: 这个和 Processor 的区别是什么

Author: ```ExtenderProcessor``` 定义的方法，入参和出参不一样。我是觉得因为不知道外部 processor 要做什么，所以就干脆把可能用到的参数都传给外部 processor。

	GetArtifactType(ctx context.Context, artifact *artifact.Artifact) (string, error)

	ListAdditionTypes(ctx context.Context, artifact *artifact.Artifact) ([]string, error)

	AbstractMetadata(ctx context.Context, artifact *artifact.Artifact) (*artifact.Artifact, error)

	AbstractAddition(ctx context.Context, artifact *artifact.Artifact, additionType string) (*Addition, error)
```

```Registry``` need to support to store both ```Processor``` and ```ExtenderProcessor```

```
var (
	// Registry for registered artifact processors
	Registry = map[string]interface{}
)
```

> When the four functions are invoked, it need to judge whether it is a ```Processor``` or ```ExtenderProcessor```

Reviewer: 这个是想表达什么

Author: 就是要要改 harbor 的逻辑，哪里调用了这 4 个方法，要判断一下是哪个 processor，因为参数不一样。

#### Use remote web service as processor

For remote processor, the four functions defined in ```Processor``` interface can be abstract to four APIs.The API route can be defined by users. It may not meet the RESTful style.

```
POST /getartifacttype
POST /listadditiontypes
POST /abstractmetadata
POST /abstractaddtion
```

Harbor core service call extender processor like this way.

```
POST extenderProcessorUrl/getartifacttype
POST extenderProcessorUrl/listadditiontypes
POST extenderProcessorUrl/abstractmetadata
POST extenderProcessorUrl/abstractaddtion
```

We can use a yaml file named ```extender_processors.yaml``` mount in core service to register ```extender processor``` when core service start.

```

> extenderProcessors:

Reviewer: 这个如何注册到 Registry 里去

Author: 现在的想法是用一个 yaml 文件，放到指定目录，core 启动的时候，去读取，然后注册。
还有一种方式，core 暴露一个接口让 ```extenderProcessor``` 注册。感觉都差不多。
